tools/torch_quant/torch_quant/graph.py

```diff
@@ -0,0 +1,234 @@
+                continue
+            if any(node.target == m for m in method_names):
+                yield node
+
+    def get_attr(self, full_path: str) -> Any:
+        parent, name = _locate_parent(self.gm, full_path)
+        return getattr(parent, name)
+
+    # TODO(litan.ls): make sure not override existing module?
+    def add_module(self, full_path: str, module: nn.Module) -> None:
+        _add_module(self.gm, full_path, module)
+        _add_module(self.root, full_path, module)
+        self.modules[full_path] = module
+
+    def replace_module(self, full_path: str, module: nn.Module) -> None:
+        _add_module(self.gm, full_path, module)
+        self.modules[full_path] = module
+
+    def register_buffer(self, full_path: str, value: torch.Tensor) -> None:
+        cloned = value.clone().detach()
+        _register_buffer(self.gm, full_path, cloned)
+        _register_buffer(self.root, full_path, cloned)
+
+    def get_or_create_module(self, full_path: str, constructor: Callable[[], nn.Module]) -> nn.Module:
+        for n, m in self.root.named_modules():
+            if n == full_path:
+                _add_module(self.gm, full_path, m)
+                return m
+
+        m = constructor()
+        self.add_module(full_path, m)
+        return m
+
+
+GraphModPass = Callable[[GraphModContext], None]
+
+
+def modify_graph(gm: GraphModule, root: nn.Module, passes: Iterable[GraphModPass]) -> None:
+    ctx = GraphModContext(gm=gm, root=root)
+    for p in passes:
+        p(ctx)
+
+
+# Some basic and generic graph modification passes:
+
+# set QConfig to quantizable modules so we can reuse nn.intrinsic/qat modules
+# TODO(litan.ls): support other observer type and dtype
+def set_qconfig(ctx: GraphModContext) -> None:
+    for node in ctx.nodes_by_module_type(QUANTIZABLE_MODULE_TYPES):
+        m = ctx.modules.get(node.target)
+        m.qconfig = QConfig(activation=None, weight=partial(
+            MinMaxObserver, dtype=torch.qint8))
+
+
+def insert_act_observer(ctx: GraphModContext) -> None:
+    # key: activation node to be observed, value: consumers of observed activation
+    # note that not all consumers of original activation need consum observed activation.
+    act_nodes: Dict[Node, Set[Node]] = defaultdict(set)
+    for node in ctx.nodes_by_module_type(QUANTIZABLE_MODULE_TYPES):
+        for arg in node.args:
+            if isinstance(arg, Node):
+                act_nodes[arg].add(node)
+        act_nodes[node].update(node.users)
+    for act in act_nodes:
+        # TODO(litan.ls): act.op == call_method
+        if act.op == 'call_function':
+            ob_path = f'{act.name}_ob'
+        else:
+            ob_path = f'{act.target}_ob'
+
+        ob_ctr = partial(MinMaxObserver, dtype=torch.quint8)
+        _ = ctx.get_or_create_module(ob_path, ob_ctr)
+        with ctx.gm.graph.inserting_after(act):
+            ob_node = ctx.gm.graph.call_module(ob_path, (act, ))
+        act.replace_all_uses_with(
```

Comment:
If enable amp, replace all uses may not be suitable. For example, given a module:
```
x  -> conv1 ->
  \  
    - -> conv2 ->
```
When only conv1 needs to be quantized, we should not change the input of conv2

Reply:
`delete_user_cb` is used to determine if the user should be switched to use new node. For example:
```
x -> conv -> 
  \ 
   - -> some_unquantizable_op ->
```
`act_nodes[x]` will only contain `conv` node, and `some_unquantizable_op` node will still connect to original `x` node after this `insert_act_observer` pass.

Reply:
However, amp could be a different story. I haven't thought about how to implement amp yet  🙃
If amp is determined before this `insert_act_observer` pass, then there should be no problem. Otherwise, we may need to handle graph re-write for amp specifically.